src/Disciplina/Educator/SizedMerkleTree.hs

```diff
@@ -0,0 +1,237 @@
+
+toLazyByteString :: Builder -> LBS.ByteString
+toLazyByteString = Builder.toLazyByteStringWith (Builder.safeStrategy 1024 4096) mempty
+
+mkLeaf :: HasHash a => (LeafIndex, a) -> SizedMerkleNode a
+mkLeaf (i, a) =
+    SizedMerkleLeaf
+    { mVal  = a
+    , mIndex = i
+    , mRoot = SizedMerkleRoot (unsafeHash a) -- unsafeHash since we need to hash to ByteString
+                              1 -- size of leaf node is 1
+    }
+
+mkBranch :: SizedMerkleNode a -> SizedMerkleNode a -> SizedMerkleNode a
+mkBranch l r =
+    SizedMerkleBranch
+    { mLeft  = l
+    , mRight = r
+    , mRoot  = mkBranchRootHash (mRoot l) (mRoot r)
+    }
+
+mkBranchRootHash :: SizedMerkleRoot a -- ^ left merkle root
+                 -> SizedMerkleRoot a -- ^ right merkle root
+                 -> SizedMerkleRoot a
+mkBranchRootHash (SizedMerkleRoot (AbstractHash hl) sl)
+                 (SizedMerkleRoot (AbstractHash hr) sr)
+  = SizedMerkleRoot
+     (hash $ toLazyByteString $ mconcat
+        [ byteString (one sl)
+        , byteString (one sr)
+        , byteString (convert hl)
+        , byteString (convert hr) ])
+     (sl + sr)
+
+-- | Smart constructor for 'SizedMerkleTree'.
+mkSizedMerkleTree :: (HasHash a, Foldable t) => t a -> SizedMerkleTree a
+mkSizedMerkleTree x = mkSizedMerkleTree' (zip [0,1..] (toList x))
+
+mkSizedMerkleTree' :: HasHash a => [(LeafIndex, a)] -> SizedMerkleTree a
+mkSizedMerkleTree' [] = SizedMerkleEmpty
+mkSizedMerkleTree' ls = SizedMerkleTree (go lsLen ls)
+  where
+    lsLen = Universum.length ls
+    go :: HasHash a => Int -> [(LeafIndex, a)] -> SizedMerkleNode a
+    go _  [x] = mkLeaf x
+    go len xs = mkBranch (go i l) (go (len - i) r)
+      where
+        i = powerOfTwo len
+        (l, r) = splitAt i xs
+
+-- | Returns root of merkle tree.
+getSizedMerkleRoot :: SizedMerkleTree a -> SizedMerkleRoot a
+getSizedMerkleRoot SizedMerkleEmpty    = emptyHash
+getSizedMerkleRoot (SizedMerkleTree x) = mRoot x
+
+emptyHash :: SizedMerkleRoot a
+emptyHash = SizedMerkleRoot (hash mempty) 0
+
+-- | Return the largest power of two such that it's smaller than X.
+--
+-- >>> powerOfTwo 64
+-- 32
+-- >>> powerOfTwo 65
+-- 64
+powerOfTwo :: (Bits a, Num a) => a -> a
+powerOfTwo n
+    | n .&. (n - 1) == 0 = n `shiftR` 1
+    | otherwise = go n
+ where
+    {- “x .&. (x - 1)” clears the least significant bit:
+           ↓
+       01101000     x
+       01100111     x - 1
+       --------
+       01100000     x .&. (x - 1)
+     -}
+    go w = if w .&. (w - 1) == 0 then w else go (w .&. (w - 1))
+
+data MerkleProof a =
+  MerkleProof { getMerkleProof :: [ProofElem a]  -- ^ list of proof elements
+              , getMerkleProofRoot :: Maybe (SizedMerkleRoot a) -- ^ leaf root proof is
+                                                                -- constructed for
+              }
+  deriving (Show, Eq, Ord, Generic, Serialise)
+
+data ProofElem a = ProofElem
```

Comment:
Why do you have separate structure to represent proof, instead of reusing the tree you described above (in that case, you should add node carrying only hash)? Do you plan to write `insert`/`delete` implementation for this new proof datatype?